### PR TITLE
[Prelim] A dex mini-probabilistic library (Try 2)

### PR DESCRIPTION
**Pull request by srush**
This is a little library for discrete variable probabilistic modeling. The main idea is to provide really simple combinators over tables  for building up the log-evidence, and then utilize auto-diff to compute the underlying marginal probabilities. There are combinators for conditioning, independent, construction, and observations. Examples for Bayes nets, dice rolling, and HMMs. The same code can be used for exact inference or sampling. 

(I originally was reading a bunch about PPLs and Haskell style monadic bayesian inference, but I am gradually feeling like the monadic style is rarely right in Dex. It leads to lots of functions, when most of the time I really want to be working with tables. I also have trouble reasoning about the efficiency. )

Example code:

```haskell
def wet ((rain_var, sprinkler_var, wet_var):
            (DVar Rain & DVar Sprinkler & DVar Binary)) :
        Probs Binary =
    rain  = sample rain_var $ bernoulli 0.2
    sprinkler_rain = cond_sample sprinkler_var [bernoulli 0.4, bernoulli 0.01] rain
    trans = for (s, r). [[bernoulli 0.0, bernoulli 0.8],
                               [bernoulli 0.9, bernoulli 0.99]].s.r                         
    sample wet_var $ trans <| sprinkler_rain

-- Obs injected through random vars
wet ({|obs=true|}, {|obs=true|}, {|obs=true|})

-- Calls grad
(rain, sprinkle) = marginalize (\ (v2, v3). wet (lat v2, lat v3, var) ?? [False, True])

-- Keys injected through random vars. 
[k1, k2] = splitKey $ newKey 0
wet ({|sample=(one, k1)|}, {|sample=(one,k2)|}, var) 
```

### CODE REVIEW

File: examples/probability.dx [hunk 1]
```diff
@@ -0,0 +1,297 @@
+
+' # Mini-PPL for Discrete variables in Dex
+
+' This document is an example for creating a small PPL in Dex. This is
+inspired by PPL's written for Haskell that utilize monadic
+represtations of the chain rule for representing discrete
+variables. Notably though in Dex we aim for a non-monadic style that
+primarily uses tables and autodifferentiation.
```
Comment (apaszke) [inline examples/probability.dx:8]: nit: autodifferentiation -> "automatic differentiation" or "autodiff". The current form reads a bit weird to me?

File: examples/probability.dx [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+' Construct a tuple from independent random variables -  $p(a, b) = p(a) p(b)$
+
+def (,,) (adist : Probs a) (bdist : Probs b) : Probs (a & b) =
+    (AsProbsTable a) = adist
+    (AsProbsTable b) = bdist 
+    AsProbsTable for (i,j). a.i * b.j
+
+' Apply the markov rule -  $p(a) = \sum_b p(a | b) p(b)$
+
+def (<|) (cond_dist : CDist a b) (prior : Probs b) : Probs a =
```
Comment (apaszke) [inline examples/probability.dx:76]: I'm not sure if our parser supports it right now, but it would be kind of cool if you could write "a `ConditionedOn` b" and get a conditional distribution. Or you could use an operator for the type constructor and write something like `a || b`.

File: examples/probability.dx [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+               (AsProbsTable c) = cond_dist.i
+               b.i * c.j
+
+-- ' Apply the chain rule -  $p(a, b) = p(a | b) p(b)$
+
+-- def chain (cond_dist : CDist a b) (prior : Probs b) : Probs (a & b) =
+--     (AsProbsTable b) = prior
+--     AsProbsTable $ for (j,i).
+--                (AsProbsTable c) = cond_dist.i
+--                b.i * c.j
```
Comment (apaszke) [inline examples/probability.dx:88]: Why is this commented out?

File: examples/probability.dx [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+
+
+' Observe a variable 
+
+def obs (x: b) (zdist : Probs a) (x_z : CDist b a) : Probs a =
+    (AsProbsTable z) = zdist
+    AsProbsTable $ for i.
+        (AsProbsTable x') = x_z.i
+        z.i .* x'.x
```
Comment (apaszke) [inline examples/probability.dx:98]: Is that meant to be an implementation of Bayes rule? Can you please add an explanation?

File: examples/probability.dx [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+' ## Random Variables
+
+' We consider random variables of three different forms.
+  Latent, observed, sampled.
+
+' 
+  * Latent variables - Sum out explicitly
+  * Observed variables - Assume observed
+  * Sampled variables - Sample out in inference
```
Comment (apaszke) [inline examples/probability.dx:109]: It would be great if you could elaborate this a little. I'm finding it a little hard to follow

File: examples/probability.dx [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+
+' For latent and sampled variables `a=>Float` should be all 1.0s.
+  This vector is used to compute marginals through autodiff.
```
Comment (apaszke) [inline examples/probability.dx:113]: Why so?

File: examples/probability.dx [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+           z' = sample hidden_vars.i $ transition <| (get zref)
+
+           -- Factor in evidence
+           zref := obs x.i z' emission    
+
+' Compute an HMM over a simple sequence
+
+show $ hmm (for i:(Fin 5). var, var) (for i. (0@_)) (arb k1) (arb k2)
+
+' Do marginalization using a autograd (Currently doesn't work yet)
```
Comment (apaszke) [inline examples/probability.dx:293]: What's the problem with that? Can you please open an issue?

**Review by apaszke — COMMENTED** [review]
Wohoo this is super cool! I'm not proficient in PPLs, so I couldn't follow the details from just the cursory look I took, but I've left some comments that might be useful.

Sorry for the delay. We've been in paper writing mode recently and so we've let GitHub slip. We should be much more active from now on (and hopefully we should have an updated of the workshop paper on arXiv soon!).

File: examples/probability.dx (hunk 7 above)
Comment (srush) [reply examples/probability.dx:293]: Will do. When there are datatypes in my auto-differentiation path (not the variables I am differentiating with respect to), I often get run-time errors. I will make some sample cases.

File: examples/probability.dx (hunk 5 above)
Comment (srush) [reply examples/probability.dx:109]: Yeah, let me write this up more formally

**Comment by apaszke** [comment]
Looks like the branch has gotten a bit messed up 😕

**Comment by srush** [comment]
Oh sorry meant to kill this PR. Going to try this again one more time, and make it really clean.
